### PR TITLE
Register

### PR DESCRIPTION
**Pull request by Yakitrak**
Adding FE register route and component using Formik and Material UI
Including Merge changes from User branch from @MidgetJake

### CODE REVIEW

File: client/src/Common/Helpers/User.tsx
Comment (BenRGH) [inline client/src/Common/Helpers/User.tsx:4]: space between third party imports and local

File: client/src/Common/Helpers/User.tsx
Comment (BenRGH) [inline client/src/Common/Helpers/User.tsx:24]: Should use array.some() here instead on the object as an array, rather than a for loop

File: client/src/Common/Helpers/User.tsx
Comment (BenRGH) [inline client/src/Common/Helpers/User.tsx:44]: .then and .catch should be on the same indentation level

File: client/src/Common/Interfaces/users.tsx
Comment (BenRGH) [inline client/src/Common/Interfaces/users.tsx:2]: Line space needed

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:2]: No need for this linespace

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:16]: Need linespace between third and local

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:27]: Better name

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:83]: No anon funcs in props

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:85]: .then and .catch need to be on same line indent

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:95]: Not 100% sure but I think this is another anon func issue since its in render

File: client/src/Components/Routing/UserContext.tsx
Comment (BenRGH) [inline client/src/Components/Routing/UserContext.tsx:22]: Interfaces shouldn't be defined inline

File: client/src/Components/Routing/UserContext.tsx
Comment (BenRGH) [inline client/src/Components/Routing/UserContext.tsx:31]: Probably best to define the state arg as Readonly<IState> to be safe, even though you're using a new copy anyway

File: client/src/Pages/Register/index.tsx
Comment (BenRGH) [inline client/src/Pages/Register/index.tsx:3]: Linespace between 3rd party and local

File: client/src/Pages/Register/index.tsx
Comment (BenRGH) [inline client/src/Pages/Register/index.tsx:6]: Shouldn't be linespace here

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:70]: No anon funcs in props

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [reply client/src/Components/RegisterForm/index.tsx:27]: Also this singe line for each thing is a bit messy, did prettier do it?

File: client/src/Common/Interfaces/users.tsx
Comment (BenRGH) [inline client/src/Common/Interfaces/users.tsx:31]: Could probably just use the ? here to be consistent

File: client/src/Common/Interfaces/users.tsx
Comment (BenRGH) [inline client/src/Common/Interfaces/users.tsx:26]: These two state interfaces (or at least this IState) should stay in the user context file with the rest of the state

File: client/src/Components/Routing/UserContext.tsx
Comment (BenRGH) [inline client/src/Components/Routing/UserContext.tsx:9]: Type

File: client/src/Components/Routing/UserContext.tsx
Comment (BenRGH) [reply client/src/Components/Routing/UserContext.tsx:31]: Also needs return type

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:43]: Return type

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [reply client/src/Components/RegisterForm/index.tsx:27]: Also needs type, unless the PasswordValidator package doesn't support ts/have types

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [inline client/src/Components/RegisterForm/index.tsx:20]: Needs more specific name

Comment (BenRGH) [reply client/src/Components/RegisterForm/index.tsx:20]: And it's an interface so needs to start with I

File: client/src/Common/Helpers/User.tsx
Comment (MidgetJake) [reply client/src/Common/Helpers/User.tsx:24]: Out of curiosity, why?

If you want the best performance, use something like:
```js
for (let i = 0, keys = Object.keys(errors); i < keys.length; i++) {
    ...
}
```

I'd be more worried about why there are now 2 `Object.keys()` calls. With one being called every single iteration

Comment (BenRGH) [reply client/src/Common/Helpers/User.tsx:24]: I suggest the .some() since this whole for block is only doing what could be done on a single line. Also yea the use of Object.keys() twice is messy

**Comment by DoctorDib** [comment]
@Yakitrak - Please resolve changes for your PR to be merged

File: client/src/Common/Helpers/User.tsx
Comment (DoctorDib) [reply client/src/Common/Helpers/User.tsx:24]: ESLint was complaining about your old method @MidgetJake - this is one of many solutions to fix the errors, feel free to implement your own fixes, but make sure ESLint approves.

**Comment by DoctorDib** [comment]
image
I approve of the use of input fields and I will be changing my form around your style, however the structure of it needs changing before we merge this... As seen in the image attached, it looks like it's only for mobile... as discussed on your Trello card https://trello.com/c/lDZVSgta/60-page-user-registration - please try to follow this style.

File: client/src/Components/RegisterForm/index.tsx
Comment (DoctorDib) [inline client/src/Components/RegisterForm/index.tsx:76]: Personally I am not a fan of an error message saying "between 8 to 100 characters" since you don't usually see the "to 100", but it is personal opinion, feel free to ignore if other disapprove

File: client/src/Pages/Register/index.tsx
Comment (DoctorDib) [inline client/src/Pages/Register/index.tsx:22]: Personally I would avoid using in lint style if possible, which is a reminder that we should go through the code and remove any existing in style props.

File: client/src/Components/Routing/UserContext.tsx
Comment (DoctorDib) [inline client/src/Components/Routing/UserContext.tsx:52]: Prop to top 👍 
In other words, Ben pointed out that the best practice is having props at the top of the function.

File: client/src/Components/Routing/UserContext.tsx
Comment (DoctorDib) [inline client/src/Components/Routing/UserContext.tsx:9]: My be worth moving this into the "Defaults" file
/Common/Helpers/Defaults.tsx

File: client/src/Components/RegisterForm/index.tsx
Comment (DoctorDib) [reply client/src/Components/RegisterForm/index.tsx:95]: Yeah, I have been working on Formik on my other branch, I recommend using the hook version called useFormik.
https://formik.org/docs/api/useFormik

useFormik will avoid using anon functions.

File: client/src/Components/RegisterForm/index.tsx
Comment (DoctorDib) [inline client/src/Components/RegisterForm/index.tsx:62]: As stated in another comment if you have this in /Common/Helpers/Defualts.tsx" you can avoid using 6-7 lines.

**Review by DoctorDib — CHANGES_REQUESTED** [review]
As discussed in a comment, please can you change the structure of the register page.

File: client/src/Components/RegisterForm/index.tsx
Comment (BenRGH) [reply client/src/Components/RegisterForm/index.tsx:76]: Yea normally it'd just say 'At least 8' and then show an error if it's too long

File: client/src/Components/Routing/UserContext.tsx
Comment (BenRGH) [reply client/src/Components/Routing/UserContext.tsx:9]: This is the one case where I'd disagree, initial state is normally kept with the state since its not a default so much as it is the starting point for state

Comment (DoctorDib) [reply client/src/Components/Routing/UserContext.tsx:9]: Fair enough, some of our code use the method I suggested :s

Comment (BenRGH) [reply client/src/Components/Routing/UserContext.tsx:9]: Yea normally its better to keep the defaults in one place like we have

Comment (DoctorDib) [reply client/src/Components/Routing/UserContext.tsx:9]: Not if you want to use that default in multiple locations, and by that point, may as well keep it all consistent

**Comment by DoctorDib** [comment]
@Yakitrak - Can we get an ETA on this please, this is holding up @BenRGH branch.

**Comment by DoctorDib** [comment]
@BenRGH - @Yakitrak  has pulled out of the project, I am going to go through this PR later and fix it up

**Comment by DoctorDib** [comment]
Remaking on a new branch https://github.com/DoctorDib/BikeTheft/tree/new-register-page

**Pull request closed without merging**
